Add editable headings for cart and checkout order summary subtotal

samueljseay opened the pull request:
### Changes proposed in this Pull Request:

Partially addresses https://github.com/woocommerce/woocommerce/issues/52043

In https://github.com/woocommerce/woocommerce/pull/52892 I implemented editable headings via entities, but overall didn't like the experience or the needless added complexity.

This takes a more traditional route with block attributes and the experience feels much better. Also the last PR was much too large so I've opted to break it into smaller PR's to address each heading one at a time. Easier for reviewing and testing.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

For cart and checkout:

1. Edit the page, find the subtotal in the order summary, click on the text to edit it
2. Smoke test various editing scenarios (empty out all the text, save empty, etc)
3. Save the changes
4. Smoke test the changes display as expected on the frontend.

Test translations:

1. Have translations installed for Nederlands (but stay in English for now)
2. Edit the page find the subtotal, edit it to anything and save
3. Now go back and edit the page again setting the heading to the english default `Subtotal`
4. Now swap to Nederlands language in WordPress
5. Now go to front-end and see that the heading should be translated to `Subtotaal`

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->
Allow order summary subtotal heading to be edited in both cart and checkout blocks.
</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

samueljseay committed:
Add editable headings for cart and checkout order summary subtotal

Commit:
Add changefile(s) from automation for the following project(s): woocommerce-blocks

nielslange commented on plugins/woocommerce-blocks/assets/js/blocks/cart/inner-blocks/cart-order-summary-subtotal/frontend.tsx:
```diff
@@ -1,6 +1,13 @@
 
-export default Block;
+const Frontend = ( { className = '', heading }: BlockAttributes ) => {
+	const headingElement = heading ?? DEFAULT_HEADING;
```
```suggestion
	const headingElement = heading || DEFAULT_HEADING;
```


Instead of using the _nullish coalescing operator_ (`??`), we should use the _logical OR operator_ (`||`). This is because the _nullish coalescing operator_ does not treat empty strings as falsy. If a merchant clears the headline in the editor, the frontend will display no headline. Using the _logical OR operator_ ensures that an empty string will fall back to the default headline.

## Nullish coalescing operator

<img width="2560" alt="Screenshot 2024-12-09 at 15 38 26" src="https://github.com/user-attachments/assets/5e5b0e45-4bb1-469c-afa6-393aea73109b">


## Logical OR operator

<img width="2560" alt="Screenshot 2024-12-09 at 15 36 16" src="https://github.com/user-attachments/assets/3a5e7a41-cc91-47e6-b20b-f205d414379e">

samueljseay replied:
@nielslange this was intentional, I was pretty cognizant of why I was using `??` here because I set the block json values to `null` as default, which is why I check for them using `??` here. This allows displaying no value in the field. I guess the core issue is, I'm not sure we agree on what the best behaviour should be here, I couldn't see a reason to disallow empty text. For example if you edit CART TOTALS right now, you certainly can set it to an empty string. That said I'd be ok if we enforced that text must be here, maybe that last point we could ask the team for input on.

nielslange replied:
Thanks for clarifying your intention to use the _nullish coalescing operator_. I appreciate the flexibility it offers to display no value. However, I'm concerned about maintaining a consistent user experience across our platform.

Typically, what merchants see in the page editor, including placeholders, is what they expect to see on the frontend. We currently ensure this consistency with the `Proceed to Checkout` and the `Place Order` button label.

Allowing merchants to leave labels empty could introduce a new behavior that might lead to confusion, especially if a merchant inadvertently overlooks an empty heading on the frontend due to seeing the placeholder in the page editor.

Could you elaborate on why a merchant would want to have one or more headings intentionally left empty on the frontend? I’m also concerned about potential legal issues, for instance, if it’s unclear which costs correspond to which items, as illustrated in this screenshot where headings like `Subtotal`, `Fees`, `Delivery`, and `Total` are removed:

<details>
<summary>Screenshot</summary>

![Screenshot showing a checkout page with missing headings](https://github.com/user-attachments/assets/f4f640da-bcd8-485c-a212-c5d7efea8c58)


<details>

samueljseay replied:
> Could you elaborate on why a merchant would want to have one or more headings intentionally left empty on the frontend?

It's inconsistent because the headers have different purposes. but one place i could imagine it, for example the "add a coupon" menu, you may want to remove the text, and then adding an icon to open it with custom style and the text would no longer intrude on your design. CSS can help with this too, so maybe I'm thinking incorrectly about the purpose of editing text. I imagined some scenario where users might delete text to get it out of the way in favor of icons or other design elements. But if you're going that far it could be argued the tool you should be using would be CSS.

> We currently ensure this consistency with the Proceed to Checkout and the Place Order button label.

One thing that was tripping me up is we allow empty values for the `CART TOTALS` heading. So we're obviously not too consistent across the codebase. But those that you mention are more recent and were done intentionally, so yeah maybe we should stick to that approach across the board. You did actually mention that to me in our review session and I forgot about it.

> Allowing merchants to leave labels empty could introduce a new behavior that might lead to confusion, especially if a merchant inadvertently overlooks an empty heading on the frontend due to seeing the placeholder in the page editor.

Yes I see your point about the placeholder. because it's subtle they won't maybe notice it's empty by accident. That is a compelling reason not to allow it. I'm torn on this one so maybe it's just safer not to allow it as you say. 👍🏻

nielslange commented on plugins/woocommerce-blocks/assets/js/blocks/checkout/inner-blocks/checkout-order-summary-subtotal/block.tsx:
```diff
@@ -5,13 +5,26 @@ import { Subtotal, TotalsWrapper } from '@woocommerce/blocks-components';
-const Block = ( { className = '' }: { className?: string } ): JSX.Element => {
+export type BlockAttributes = {
+	className?: string;
+	heading?: string;
```
```suggestion
	heading: string;
```

Shouldn't the heading attribute be required? It's either the custom heading of the merchant or the placeholder, isn't it?

samueljseay replied:
Dang thanks for spotting this, looks like I relied too much on the AI to generate these type defs, whoops should be `string | null`. Will fix.

nielslange replied:
As for my understanding, should it only be `string | null` when we allow displaying an empty string on the frontend? This goes back to my question in https://github.com/woocommerce/woocommerce/pull/53544#discussion_r1875943473. Or should this definition be  `string | null` independent of us allowing merchants to display empty strings?

samueljseay replied:
@nielslange in light of changing the UX experience we will change this for sure. but if the JSON definition had a default of `null` then this would be `string | null`. Maybe that's obvious but just pointing that out. The type definition will match how the attribute is defined.

nielslange commented on plugins/woocommerce-blocks/assets/js/blocks/cart/inner-blocks/cart-order-summary-subtotal/block.tsx:
```diff
@@ -5,13 +5,29 @@ import { Subtotal, TotalsWrapper } from '@woocommerce/blocks-components';
 
-const Block = ( { className = '' }: { className?: string } ): JSX.Element => {
+export type BlockAttributes = {
+	className: string;
```
```suggestion
	className?: string | undefined;
```

Shouldn't this definition also check against an optional element that can be `undefined`?

samueljseay replied:
I don't think it needs to be `className?: string | undefined`

But you're right that it's probably not string. I copied the type def from here: https://github.com/woocommerce/woocommerce/pull/53544/files/323cc868b3139fdea19f33153c43cc8656380552#diff-090b2440beb63e4a22de7aac648a50ee7cb9cc3023d39e4c9033cca4dab09e60L14-L16 because that's whats passed down.

I have actually seen className typed many ways throughout the code base.

in JSX it's a little nuanced:

lets say that the `Edit` block has `undefined` for the `className` attribute like [here](https://github.com/woocommerce/woocommerce/pull/53544/files/323cc868b3139fdea19f33153c43cc8656380552#diff-090b2440beb63e4a22de7aac648a50ee7cb9cc3023d39e4c9033cca4dab09e60R18) and we always render the Block like: `<Block className={className} ... />` like [here](https://github.com/woocommerce/woocommerce/pull/53544/files/323cc868b3139fdea19f33153c43cc8656380552#diff-090b2440beb63e4a22de7aac648a50ee7cb9cc3023d39e4c9033cca4dab09e60R41) then using `?` is incorrect, the property will always be present, but assuming the `className` in edit could be undefined then the correct type would be `className: string | undefined`

That's because we always pass the property key by doing `className={ className }`. if we did something different like (naive example):

```
let className: string | undefined;

const props = {
  foo: 'bar'
}

if ( className ) { props.className = className }
 
 <Block {...props} /> 
 ```
then the type could indeed be `className?: string | undefined`

samueljseay replied:
For a while now I'd been asking myself, actually is this className type correct at all, we tend to type it incorrectly and inconsistently, but I just got to the bottom of it.

`BlockEditProps` is a convenience type for the props passed to edit. it extends `BlockSaveProps` which has `className` typed as string. Which makes sense. Actually these **should all be string**. In that case I would assume that the frontend block should also have className as `string` too.

nielslange replied:
> lets say that the Edit block has undefined for the className attribute like [here](https://github.com/woocommerce/woocommerce/pull/53544/files/323cc868b3139fdea19f33153c43cc8656380552#diff-090b2440beb63e4a22de7aac648a50ee7cb9cc3023d39e4c9033cca4dab09e60R18) and we always render the Block like: <Block className={className} ... /> like [here](https://github.com/woocommerce/woocommerce/pull/53544/files/323cc868b3139fdea19f33153c43cc8656380552#diff-090b2440beb63e4a22de7aac648a50ee7cb9cc3023d39e4c9033cca4dab09e60R41) then using ? is incorrect, the property will always be present, but assuming the className in edit could be undefined then the correct type would be className: string | undefined

That's what I looked up. In `edit.tsx`, I noticed that `className` is `undefined`.

samueljseay replied:
@nielslange Yep I was pointing out the difference between `className?: string | undefined` and `className: string | undefined`. I was trying to explain why `className?` is not appropriate here.

nielslange commented on plugins/woocommerce-blocks/assets/js/blocks/cart/inner-blocks/cart-order-summary-subtotal/block.tsx:
```diff
@@ -5,13 +5,29 @@ import { Subtotal, TotalsWrapper } from '@woocommerce/blocks-components';
-const Block = ( { className = '' }: { className?: string } ): JSX.Element => {
+export type BlockAttributes = {
+	className: string;
+	heading: string | null;
```
```suggestion
	heading: string;
```

Shouldn't this element be always available? Either it's the custom heading of the merchant or the placeholder text.

samueljseay replied:
@nielslange no, this is the attribute as it's set in the block.json which has a default of null therefore can be string | null

nielslange replied:
I assume this also depends on the question I raised in https://github.com/woocommerce/woocommerce/pull/53544#discussion_r1875964453, doesn't it?

samueljseay replied:
@nielslange I guess the point is from a review perspective the implementation of this type is not incorrect and I wanted to make that clear. if we change the code then yes the block def will change but what does the UX change have to do with that here? If I change the code then naturally this type definition will change.

samueljseay replied:
@nielslange I don't want to split hairs but the implementation of the type was correct. We implement types as we expect the type to be. When we change the UX we will change the code because the block.json attribute might have default to string to achieve that. In that case yes we'll change the definition 😛 but we don't need to go over this now. We'll change the implementation, the code will change, then you'll re-review.

nielslange commented on plugins/woocommerce-blocks/packages/components/totals/subtotal/index.tsx:
This file still had the following entry at the top, which we should remove:

```js
import { __ } from '@wordpress/i18n';
```

nielslange reviewed (changes requested):
Thanks for working on this, @samueljseay. While testing the PR, I noticed that when the merchant empties the headings in the page editor, the frontend does not display the placeholder texts. That's because in the frontend, we're not using the `<RichText>` component, where we can define a placeholder. I left feedback below with a suggested solution. I also noticed a few TS definitions that were unclear to me. Please let me know if you have any questions regarding my review.

samueljseay committed:
Do not allow empty string but keep support for translated strings.

samueljseay commented:
@nielslange Could you take another look at this, in 2 steps:

1. Could you UI test the new behaviour introduced in https://github.com/woocommerce/woocommerce/pull/53544/commits/76bdeb7d0d15fd781881dd1c2e88569cac752108

Notable:

a. if user empties out string, it shows placeholder. they save, it will save but it will not display empty string on frontend it will display placeholder. Also if they refresh the editor the placeholder will be gone and the string will be reverted to the default. For me this felt really nice, but keen to hear if you agree.

b. lets say the user edits the text, then edits it back to the default heading. In this case we actually under the hood clear the attribute they saved, so we can revert to the translated header. if we store a fixed string exactly the same as the default then users in different languages will see an untranslated string. i believe this is undesirable which is why i implemented this behaviour but also keen to hear your thoughts, this way store owners cannot accidentally break translation and from user perspective it still feels correct.



2. Before reviewing code can you let me know if you have any problems with the behaviours I have implemented
3. Code review after that

samueljseay commented on plugins/woocommerce-blocks/assets/js/blocks/cart/inner-blocks/cart-order-summary-subtotal/edit.tsx:
```diff
@@ -1,26 +1,54 @@
+
+	const onChangeCallback = useCallback(
+		( value: string ) => {
+			setHeadingText( value );
```
To support reverting to translated string we maintain a piece of state to show the rich text edited value in the editor, this is so that we can display the up to date edited value to the user, even if maybe we reset the attribute under the hood, such as we do [here](https://github.com/woocommerce/woocommerce/pull/53544/files#diff-090b2440beb63e4a22de7aac648a50ee7cb9cc3023d39e4c9033cca4dab09e60R30)

nielslange replied:
Nice way of ensuring that the translation remains working when a merchant resets the heading. Well done, @samueljseay. 👏

samueljseay committed:
Improve wording of comments

samueljseay committed:
Fix lint issues

samueljseay committed:
Merge branch 'trunk' into dev/editable-summary-subtotal-heading

nielslange reviewed (approved):
Thanks for working on this, @samueljseay. I'm pre-approving the PR as it works as expected. As for the testing instructions, do you think we should go more into detail, so that external testers know what to look out for, e.g. testing the translations after the heading got reset?
